plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -73,6 +73,11 @@ type predicateMetadata struct {
 	podRequest                *schedulercache.Resource
 	podPorts                  map[int]bool
 	matchingAntiAffinityTerms []matchingPodAntiAffinityTerm
+
+	// CheckServiceAffinity: pods cache, otherwise this is a large query, will slow things down
+	// when there are 100s of services/pods.
+	lock        sync.Mutex
```

Comment:
I'm really not a fan of this approach. I think that really instead of it, we should pass something like names of all predicates to "PredicateMetada" function:
https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/generic_scheduler.go#L162

and based on the fact if the ServiceAffinity is registered or not precompute servicePods.